### DISCUSSION TITLE
Workaround for 5.3 breaking change with ISession.Persist() (no id returned)

### DISCUSSION THREAD

**Discussion by Tony322** [nhibernate/nhibernate-core#3234] [Q&A]
Since v.5.3 there was a breaking change:
"ISession.Persist will no more trigger immediate generation of identifier."
This have caused major issues for us, since we have lots of code that relies on the Id being present on the object after session.Merge(entity) has been called within a uncommited transaction. 

Here is an issue posted by some other user explaining the issue:
https://github.com/nhibernate/nhibernate-core/issues/2632

What is the correct way of making it work like it used to (or equivalent)? Forcing flushes after merge? Using saveOrUpdate intead of merge?
This has made us really uneasy upgrading since this seem to affect our whole system and lots of business logic has broken down due to the missing id's.

Any help is much appreciated.

**Comment by fredericDelaporte** [comment]
> Forcing flushes after merge

That is the way to ensure IDs generated on DB's side are generated, with minimal impact on how the application work. (It would be slightly more optimal to do it only right before it gets actually needed, but depending on how your application works, it could be hard to know if it is needed or not, and where. Of course if the merged entities do not have DB's side generated IDs, the flush is then not needed.)

`SaveOrUpdate` works quite differently, causing the provided object to be associated with the session, which `Merge` does not do. It may or may not be a trouble for your application.

**Comment by bahusoid** [comment]
You can restore old `Merge` behavior using custom `IMergeEventListener`:

```C#
public class ForceSaveIdMergeEventListener : DefaultMergeEventListener
{
    protected override object SaveWithGeneratedId(object entity, string entityName, object anything,
        IEventSource source,
        bool requiresImmediateIdAccess)
    {
        return base.SaveWithGeneratedId(entity, entityName, anything, source, requiresImmediateIdAccess: true);
    }
}
```

And to use it:

```C#
//Somewhere before session factory is created:
NHibernate.Cfg.Configuration configuration = ...
configuration.EventListeners.MergeEventListeners = new[]{new ForceSaveIdMergeEventListener()};
```

Or via xml:
```xml
<hibernate-configuration>
    <session-factory>
        ...
        <event type="merge">
            <listener class="ForceSaveIdMergeEventListener"/>
        </event>
    </session-factory>
</hibernate-configuration>
```

Similar fix can be applied for `Persist` too if really needed (though it's better to replace usages with `session.Save`)

**Comment by Tony322** [comment]
Thank you very much for your suggestions. We went with the Merge and force flush route since we didn't feel comfortable overriding the default merge behaviour due to future proofing reasons. Thanks 👍👍
